### PR TITLE
feat(gov): custom voting period according to proposal content

### LINKED ISSUE #2: feat(gov): adapt voting period to proposal content
## Summary

We want the voting period of proposals to be 1 year, in order to allow any participant to vote, even for late-comers.

## Problem Definition

This is problematic for upgrade or params proposals, they should have a shorter voting period.

## Proposal

Adapt the voting period according to proposal content. 
- text proposal voting period : 1 year
- params change proposal voting period : 2 weeks
- upgrade proposal voting period: 4 weeks

### PR DESCRIPTION
**Pull request by tbruyelle**
Fix #2 

- update `types.VotingParams` : has 3 `VotingPeriod*` fields instead of one, for each proposal we want to support.
- add `keeper.GetVotingPeriod()` method to returns the correct voting period according to the proposal content type. Unsupported proposal content returns the default voting of 2 days.
- tests update and improvements

TODO:

Update https://github.com/atomone-hub/govgen-genesis/blob/main/base-genesis.json to reflect these changes in the voting params.

### CODE REVIEW

File: proto/govgen/gov/v1beta1/gov.proto [hunk 1]
```diff
@@ -164,11 +164,23 @@ message DepositParams {
...
-  google.protobuf.Duration voting_period = 1 [
+  google.protobuf.Duration voting_period_params_change = 1 [
     (gogoproto.nullable)    = false,
     (gogoproto.stdduration) = true,
-    (gogoproto.jsontag)     = "voting_period,omitempty",
-    (gogoproto.moretags)    = "yaml:\"voting_period\""
+    (gogoproto.jsontag)     = "voting_period_params_change,omitempty",
+    (gogoproto.moretags)    = "yaml:\"voting_period_params_change\""
+  ];
+  google.protobuf.Duration voting_period_upgrade = 2 [
```
Comment (giunatale) [inline proto/govgen/gov/v1beta1/gov.proto:173]:
```suggestion
  google.protobuf.Duration voting_period_software_upgrade = 2 [
```

**Review by giunatale — CHANGES_REQUESTED** [review]
We should use the same names as per the proposal types. I suggested changes in the proto but it should be reflected also in other places

Moreover,
I think a default voting period onto which other proposal types default too instead of panicking would be more suitable.

**Comment by tbruyelle** [comment]
> We should use the same names as per the proposal types. I suggested changes in the proto but it should be reflected also in other places
> 
> Moreover, I think a default voting period onto which other proposal types default too instead of panicking would be more suitable.

Good points, I addressed them.

File: x/gov/simulation/genesis.go [hunk 2]
```diff
@@ -17,12 +17,14 @@ import (
...
 const (
-	DepositParamsMinDeposit    = "deposit_params_min_deposit"
-	DepositParamsDepositPeriod = "deposit_params_deposit_period"
-	VotingParamsVotingPeriod   = "voting_params_voting_period"
-	TallyParamsQuorum          = "tally_params_quorum"
-	TallyParamsThreshold       = "tally_params_threshold"
-	TallyParamsVeto            = "tally_params_veto"
+	DepositParamsMinDeposit              = "deposit_params_min_deposit"
+	DepositParamsDepositPeriod           = "deposit_params_deposit_period"
+	VotingParamsVotingPeriodParamsChange = "voting_params_voting_period_params_change"
```
Comment (giunatale) [inline x/gov/simulation/genesis.go:22]: Probably need to do search to replace, this is still using old naming for example

File: x/gov/keeper/proposal.go [hunk 3]
```diff
@@ -184,9 +187,21 @@ func (keeper Keeper) SetProposalID(ctx sdk.Context, proposalID uint64) {
...
+func (keeper Keeper) GetVotingPeriod(ctx sdk.Context, content types.Content) time.Duration {
+	switch content.(type) {
+	case *types.TextProposal:
+		return keeper.GetVotingParams(ctx).VotingPeriodText
+	case *paramsproposal.ParameterChangeProposal:
+		return keeper.GetVotingParams(ctx).VotingPeriodParameterChange
+	case *upgradetypes.SoftwareUpgradeProposal, *upgradetypes.CancelSoftwareUpgradeProposal:
+		return keeper.GetVotingParams(ctx).VotingPeriodSoftwareUpgrade
+	}
+	return types.DefaultPeriod
```
Comment (giunatale) [inline x/gov/keeper/proposal.go:199]: I didn't mean a hard-coded default btw, but an equally changeable parameter.
`DefaultVotingPeriod` or something.

**Review by giunatale — CHANGES_REQUESTED** [review]
see comments

File: x/gov/simulation/genesis.go (hunk 2 above)
Comment (giunatale) [reply x/gov/simulation/genesis.go:22]: as well other places

**Comment by giunatale** [comment]
would be nice to squeeze in an update of the Readme.md here since we made some changes so far that aren't reported there.

File: x/gov/simulation/genesis.go (hunk 2 above)
Comment (tbruyelle) [reply x/gov/simulation/genesis.go]: indeed I did an other pass [c980a5c](https://github.com/atomone-hub/govgen/pull/16/commits/c980a5c5099770b76e7a6c6fc040399f82240510)

File: x/gov/keeper/proposal.go (hunk 3 above)
Comment (tbruyelle) [reply x/gov/keeper/proposal.go:199]: With which value ?

Comment (giunatale) [reply x/gov/keeper/proposal.go:199]: With another parameter.

types.DefaultPeriod is just a hard-coded constant. No bueno imho

**Comment by tbruyelle** [comment]
> would be nice to squeeze in an update of the Readme.md here since we made some changes so far that aren't reported there.

We can do that before merging `feat/custom-gov-module` to `master` OK?

**Comment by giunatale** [comment]
> > would be nice to squeeze in an update of the Readme.md here since we made some changes so far that aren't reported there.
> 
> We can do that before merging `feat/custom-gov-module` to `master` OK?

Perfect!

File: x/gov/keeper/proposal.go (hunk 3 above)
Comment (tbruyelle) [reply x/gov/keeper/proposal.go]: Added a new parameter `VotingParams.VotingPeriodDefault` [ee2bded](https://github.com/atomone-hub/govgen/pull/16/commits/ee2bded6b8557a6180f423e5c9df66d5e0781211)

Comment (tbruyelle) [reply x/gov/keeper/proposal.go]: Done here [ee2bded](https://github.com/atomone-hub/govgen/pull/16/commits/ee2bded6b8557a6180f423e5c9df66d5e0781211)

**Review by giunatale — APPROVED** [review]
the new param should be added to base genesis too.

LGTM

**Pull request merged**